control-plane-operator/controllers/hostedcontrolplane/etcd/reconcile.go

```diff
@@ -1,209 +1,316 @@
+					},
+				},
 			},
 		},
+	}
+	ss.Spec.Template.Labels = etcdPodSelector
+
+	ss.Spec.Template.Spec.Containers = []corev1.Container{
+		util.BuildContainer(etcdContainer(), buildEtcdContainer(p, ss.Namespace)),
+	}
+
+	ss.Spec.Template.Spec.Volumes = []corev1.Volume{
 		{
-			APIGroups: []string{
-				corev1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"pods",
-				"services",
-				"endpoints",
-				"persistentvolumeclaims",
-				"events",
-			},
-			Verbs: []string{
-				"*",
+			Name: "peer-tls",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: manifests.EtcdPeerSecret(ss.Namespace).Name,
+				},
 			},
 		},
 		{
-			APIGroups: []string{
-				appsv1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"deployments",
-			},
-			Verbs: []string{
-				"*",
+			Name: "server-tls",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: manifests.EtcdServerSecret(ss.Namespace).Name,
+				},
 			},
 		},
 		{
-			APIGroups: []string{
-				corev1.SchemeGroupVersion.Group,
-			},
-			Resources: []string{
-				"secrets",
-			},
-			Verbs: []string{
-				"get",
+			Name: "client-tls",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: manifests.EtcdClientSecret(ss.Namespace).Name,
+				},
 			},
 		},
 	}
-	return nil
-}
 
-func ReconcileOperatorRoleBinding(roleBinding *rbacv1.RoleBinding, ownerRef config.OwnerRef) error {
-	ownerRef.ApplyTo(roleBinding)
-	serviceAccount := manifests.EtcdOperatorServiceAccount(roleBinding.Namespace)
-	roleBinding.RoleRef = rbacv1.RoleRef{
-		APIGroup: rbacv1.SchemeGroupVersion.Group,
-		Kind:     "Role",
-		Name:     "etcd-operator",
-	}
-	roleBinding.Subjects = []rbacv1.Subject{
-		{
-			Kind:      "ServiceAccount",
-			APIGroup:  corev1.SchemeGroupVersion.Group,
-			Namespace: serviceAccount.Namespace,
-			Name:      serviceAccount.Name,
-		},
-	}
-	return nil
-}
+	p.DeploymentConfig.ApplyToStatefulSet(ss)
 
-var etcdOperatorDeploymentLabels = map[string]string{
-	"name":                        "etcd-operator",
-	"app":                         "etcd-operator",
-	hyperv1.ControlPlaneComponent: "etcd-operator",
+	return nil
 }
 
-func etcdOperatorContainer() *corev1.Container {
-	return &corev1.Container{
-		Name: "etcd-operator",
-	}
-}
+func buildEtcdContainer(p *EtcdParams, namespace string) func(c *corev1.Container) {
+	return func(c *corev1.Container) {
+		script := `
+/usr/bin/etcd \
+--data-dir=/var/lib/data \
+--name=${HOSTNAME} \
+--initial-advertise-peer-urls=https://${HOSTNAME}.etcd-discovery.${NAMESPACE}.svc:2380 \
+--listen-peer-urls=https://${POD_IP}:2380 \
+--listen-client-urls=https://${POD_IP}:2379,https://localhost:2379 \
+--advertise-client-urls=https://${HOSTNAME}.etcd-client.${NAMESPACE}.svc:2379 \
+--listen-metrics-urls=https://${POD_IP}:2381,https://localhost:2381 \
```

Comment:
is it a concern that monitoring will have access to the etcd client certs?

Reply:
Added a TODO about this